File: python/tests/test_p5.py
"""P5: causal-structure implications, structured version diff, archive bundle."""
import json

import pytest

import theoryforge as tf


@pytest.fixture
def panic(panic_path):
    return tf.read(panic_path)


@pytest.fixture
def mediation(fixtures_dir):
    return tf.read(fixtures_dir / "mediation-demo.theory.yaml")


# -- implications ---------------------------------------------------------------


def test_implications_cyclic_network(panic):
    imp = panic.implications()
    assert imp["acyclic"] is False
    assert imp["order"] == []
    assert imp["implications"] == []
    assert imp["n_implications"] == 0
    # the arousal <-> threat feedback loop, reported once from its lowest-index node
    assert imp["feedback_loops"] == [["c_arousal", "c_perceived_threat"]]
    assert imp["exogenous"] == []
    assert set(imp["endogenous"]) == set(imp["constructs"])


def test_implications_acyclic_chain(mediation):
    imp = mediation.implications()
    assert imp["acyclic"] is True
    assert imp["order"] == ["m_stressor", "m_appraisal", "m_anxiety"]
    assert imp["exogenous"] == ["m_stressor"]
    assert imp["endogenous"] == ["m_appraisal", "m_anxiety"]
    assert imp["feedback_loops"] == []
    # the classic mediation independence: anxiety _||_ stressor | appraisal
    assert imp["implications"] == [
        {"a": "m_anxiety", "b": "m_stressor", "given": ["m_appraisal"]}
    ]
    assert imp["n_implications"] == 1


def test_implications_ignores_non_causal_relations(weak_path):
    imp = tf.read(weak_path).implications()
    # 'associates' is not causal: both constructs are exogenous and marginally independent
    assert imp["exogenous"] == ["k_motivation", "k_drive"]
    assert imp["implications"] == [{"a": "k_motivation", "b": "k_drive", "given": []}]


def test_implications_deduplicates_edges_and_pairs():
    t = tf.new_theory("t", "T")
    t.add_construct("a", "A", "a.").add_construct("b", "B", "b.")
    t.add_proposition("p1", "a", "b", "increases")
    t.add_proposition("p2", "a", "b", "causes")  # duplicate (a, b) edge
    imp = t.implications()
    assert imp["acyclic"] is True
    # b's only non-descendant is its parent a, so no claim is emitted
    assert imp["implications"] == []


def test_implications_self_loop_is_a_cycle():
    t = tf.new_theory("t", "T")
    t.add_construct("a", "A", "a.")
    t.add_proposition("p1", "a", "a", "increases")
    imp = t.implications()
    assert imp["acyclic"] is False
    assert imp["feedback_loops"] == [["a"]]


def test_implications_matches_golden(fixtures_dir):
    for fx in sorted(fixtures_dir.glob("*.theory.yaml")):
        t = tf.read(fx)
        golden = json.loads(
            (fixtures_dir / "expected" / f"{t.id}.implications.json").read_text(encoding="utf-8")
        )
        assert t.implications() == golden


# -- diff -------------------------------------------------------------------------


def test_diff_amended_pair(fixtures_dir, panic):
    v2 = tf.read(fixtures_dir / "panic-network-2026-v2.theory.yaml")
    d = v2.diff(panic)
    assert d["prior_id"] == "panic-network-2026"
    assert d["new_id"] == "panic-network-2026-v2"
    assert d["changed_fields"] == ["maturity"]
    assert d["predictions"]["added"] == ["pred4"]
    assert d["summary"]["n_added"] == 1
    golden = json.loads(
        (fixtures_dir / "expected" / "panic-network-2026-v2.diff.json").read_text(encoding="utf-8")
    )
    assert d == golden


def test_diff_identical_theories_report_nothing(panic):
    d = panic.diff(panic)
    assert d["changed_fields"] == []
    assert d["summary"] == {"n_added": 0, "n_removed": 0, "n_modified": 0}
    for coll in ("constructs", "propositions", "predictions",
                 "auxiliary_assumptions", "alternatives"):
        assert d[coll] == {"added": [], "removed": [], "modified": []}


def test_diff_detects_added_removed_modified():
    prior = tf.new_theory("t", "T")
    prior.add_construct("a", "A", "a.").add_construct("b", "B", "b.")
    new = tf.new_theory("t", "T")
    new.add_construct("a", "A", "a (refined).").add_construct("c", "C", "c.")
    d = new.diff(prior)
    assert d["constructs"] == {"added": ["c"], "removed": ["b"], "modified": ["a"]}


def test_diff_canonicalisation_treats_singleton_as_scalar():
    from theoryforge.diff import canon
    assert canon(["p1"]) == canon("p1")
    assert canon({"x": None, "y": 1}) == canon({"y": 1.0})
    assert canon(True) == "true"
    assert canon(0.5) == "0.5"


# -- fair_export --------------------------------------------------------------------


def test_fair_export_contents(panic):
    files = panic.fair_export(authors=["Doe, Jane"])
    assert sorted(files) == ["CITATION.cff", "README.md", "dossier.md", "metadata.json"]
    assert files["README.md"].startswith("# Network theory of panic disorder\n")
    assert "- Version: v1\n" in files["README.md"]
    assert "authors:\n  - name: Doe, Jane" in files["CITATION.cff"]
    meta = json.loads(files["metadata.json"])  # hand-rendered but valid JSON
    assert meta["upload_type"] == "dataset"
    assert meta["keywords"] == ["scientific-theory", "theoryforge", "panic-network-2026"]
    # cited DOIs from evidence + alternatives, normalised, deduplicated, sorted
    assert {r["identifier"] for r in meta["related_identifiers"]} == {
        "10.1016/j.brat.2015.10.002",
        "10.1016/0005-7967(86)90011-2",
        "10.1176/ajp.146.2.148",
    }
    assert all(r["relation"] == "cites" for r in meta["related_identifiers"])
    assert files["dossier.md"] == panic.dossier()


def test_fair_export_defaults_and_empty_authors():
    t = tf.new_theory("demo-x", "A demo")
    files = t.fair_export()
    assert "authors: []" in files["CITATION.cff"]
    assert "- Version: unversioned\n" in files["README.md"]
    meta = json.loads(files["metadata.json"])
    assert meta["creators"] == []
    assert meta["related_identifiers"] == []


def test_fair_export_writes_bundle(tmp_path, panic):
    out = tmp_path / "bundle"
    files = panic.fair_export(path=out, authors=["Doe, Jane"])
    for name in files:
        assert (out / name).read_bytes().decode("utf-8") == files[name]
    reread = tf.read(out / "theory.yaml")
    assert reread.id == panic.id


def test_fair_export_matches_golden(fixtures_dir, panic):
    golden = json.loads(
        (fixtures_dir / "expected" / "panic-network-2026.fair.json").read_text(encoding="utf-8")
    )
    assert panic.fair_export(authors=["Doe, Jane"]) == golden

File: r/theoryforge/tests/testthat/test-p5.R
# P5: causal-structure implications, structured version diff, archive bundle.

test_that("tf_implications flags the panic network's feedback loop", {
  theory <- tf_read(tf_fixture_path("panic-network.theory.yaml"))
  imp <- tf_implications(theory)
  expect_false(imp$acyclic)
  expect_length(imp$order, 0L)
  expect_length(imp$implications, 0L)
  expect_equal(imp$n_implications, 0L)
  expect_equal(imp$feedback_loops,
               list(list("c_arousal", "c_perceived_threat")))
  expect_length(imp$exogenous, 0L)
  expect_setequal(unlist(imp$endogenous), unlist(imp$constructs))
})

test_that("tf_implications derives the mediation independence on an acyclic chain", {
  theory <- tf_read(tf_fixture_path("mediation-demo.theory.yaml"))
  imp <- tf_implications(theory)
  expect_true(imp$acyclic)
  expect_equal(unlist(imp$order), c("m_stressor", "m_appraisal", "m_anxiety"))
  expect_equal(unlist(imp$exogenous), "m_stressor")
  expect_length(imp$feedback_loops, 0L)
  expect_equal(imp$implications,
               list(list(a = "m_anxiety", b = "m_stressor", given = list("m_appraisal"))))
  expect_equal(imp$n_implications, 1L)
})

test_that("tf_implications ignores non-causal relations", {
  theory <- tf_read(tf_fixture_path("weak-theory.theory.yaml"))
  imp <- tf_implications(theory)
  expect_equal(unlist(imp$exogenous), c("k_motivation", "k_drive"))
  expect_equal(imp$implications,
               list(list(a = "k_motivation", b = "k_drive", given = list())))
})

test_that("tf_implications treats a self-loop as a cycle", {
  theory <- tf_theory("t", "T") |>
    tf_add_construct("a", "A", "a.") |>
    tf_add_proposition("p1", "a", "a", "increases")
  imp <- tf_implications(theory)
  expect_false(imp$acyclic)
  expect_equal(imp$feedback_loops, list(list("a")))
})

test_that("tf_implications matches the Python goldens semantically", {
  fixtures <- list.files(tf_fixtures_dir(), pattern = "\\.theory\\.yaml$", full.names = TRUE)
  expect_gt(length(fixtures), 0L)
  for (fx in fixtures) {
    theory <- tf_read(fx)
    golden <- jsonlite::fromJSON(tf_expected_path(paste0(theory$id, ".implications.json")),
                                 simplifyVector = FALSE)
    got <- jsonlite::fromJSON(
      jsonlite::toJSON(tf_implications(theory), auto_unbox = TRUE),
      simplifyVector = FALSE)
    expect_equal(got, golden, info = fx)
  }
})

test_that("tf_diff reports the amended pair's editorial record and matches the golden", {
  v1 <- tf_read(tf_fixture_path("panic-network.theory.yaml"))
  v2 <- tf_read(tf_fixture_path("panic-network-2026-v2.theory.yaml"))
  d <- tf_diff(v2, v1)
  expect_equal(d$prior_id, "panic-network-2026")
  expect_equal(d$new_id, "panic-network-2026-v2")
  expect_equal(unlist(d$changed_fields), "maturity")
  expect_equal(unlist(d$predictions$added), "pred4")
  expect_equal(d$summary$n_added, 1L)
  golden <- jsonlite::fromJSON(tf_expected_path("panic-network-2026-v2.diff.json"),
                               simplifyVector = FALSE)
  got <- jsonlite::fromJSON(jsonlite::toJSON(d, auto_unbox = TRUE), simplifyVector = FALSE)
  expect_equal(got, golden)
})

test_that("tf_diff on identical theories reports nothing", {
  theory <- tf_read(tf_fixture_path("panic-network.theory.yaml"))
  d <- tf_diff(theory, theory)
  expect_length(d$changed_fields, 0L)
  expect_equal(d$summary, list(n_added = 0L, n_removed = 0L, n_modified = 0L))
  for (coll in c("constructs", "propositions", "predictions",
                 "auxiliary_assumptions", "alternatives")) {
    expect_length(d[[coll]]$added, 0L)
    expect_length(d[[coll]]$removed, 0L)
    expect_length(d[[coll]]$modified, 0L)
  }
})

test_that("tf_diff detects added, removed, and modified elements", {
  prior <- tf_theory("t", "T") |>
    tf_add_construct("a", "A", "a.") |>
    tf_add_construct("b", "B", "b.")
  new <- tf_theory("t", "T") |>
    tf_add_construct("a", "A", "a (refined).") |>
    tf_add_construct("c", "C", "c.")
  d <- tf_diff(new, prior)
  expect_equal(unlist(d$constructs$added), "c")
  expect_equal(unlist(d$constructs$removed), "b")
  expect_equal(unlist(d$constructs$modified), "a")
})

test_that("the canonical serialisation treats a singleton as its scalar", {
  expect_identical(.tf_canon(list("p1")), .tf_canon("p1"))
  expect_identical(.tf_canon(TRUE), "true")
  expect_identical(.tf_canon(0.5), "0.5")
  expect_identical(.tf_canon(2), "2")
})

test_that("tf_fair_export renders the bundle and matches the golden byte-for-byte", {
  theory <- tf_read(tf_fixture_path("panic-network.theory.yaml"))
  files <- tf_fair_export(theory, authors = "Doe, Jane")
  expect_named(files, c("README.md", "CITATION.cff", "metadata.json", "dossier.md"))
  expect_match(files[["README.md"]], "^# Network theory of panic disorder\n")
  expect_match(files[["CITATION.cff"]], "authors:\n  - name: Doe, Jane", fixed = TRUE)
  expect_identical(files[["dossier.md"]], tf_dossier(theory))
  meta <- jsonlite::fromJSON(files[["metadata.json"]], simplifyVector = FALSE)
  expect_equal(meta$upload_type, "dataset")
  expect_equal(unlist(meta$keywords),
               c("scientific-theory", "theoryforge", "panic-network-2026"))
  ids <- vapply(meta$related_identifiers, function(r) r$identifier, character(1))
  expect_setequal(ids, c("10.1016/j.brat.2015.10.002",
                         "10.1016/0005-7967(86)90011-2",
                         "10.1176/ajp.146.2.148"))
  golden <- jsonlite::fromJSON(tf_expected_path("panic-network-2026.fair.json"),
                               simplifyVector = FALSE)
  for (name in names(golden)) {
    expect_identical(files[[name]], golden[[name]], info = name)
  }
})

test_that("tf_fair_export defaults cover the empty cases", {
  theory <- tf_theory("demo-x", "A demo")
  files <- tf_fair_export(theory)
  expect_match(files[["CITATION.cff"]], "authors: []", fixed = TRUE)
  expect_match(files[["README.md"]], "- Version: unversioned", fixed = TRUE)
  meta <- jsonlite::fromJSON(files[["metadata.json"]], simplifyVector = FALSE)
  expect_length(meta$creators, 0L)
  expect_length(meta$related_identifiers, 0L)
})

test_that("tf_fair_export writes the bundle plus theory.yaml when a path is given", {
  theory <- tf_read(tf_fixture_path("panic-network.theory.yaml"))
  out <- file.path(tempdir(), "tf-fair-bundle")
  on.exit(unlink(out, recursive = TRUE), add = TRUE)
  files <- tf_fair_export(theory, path = out, authors = "Doe, Jane")
  for (name in names(files)) {
    expect_identical(tf_read_golden(file.path(out, name)), files[[name]])
  }
  reread <- tf_read(file.path(out, "theory.yaml"))
  expect_equal(reread$id, theory$id)
})
